rvgo/test/syscall_test.go

```diff
@@ -0,0 +1,419 @@
+			// We pre-allocate a buffer for the read hint data to be copied into.
+			LastHint: make(hexutil.Bytes, fast.PageSize),
+		}
+		state.Memory.SetUnaligned(pc, syscallInsn)
+		preStatePreimageKey := state.PreimageKey
+		preStateRoot := state.Memory.MerkleRoot()
+		expectedRegisters := state.Registers
+		expectedRegisters[10] = count
+		expectedRegisters[11] = 0
+
+		oracle := staticOracle(t, preimageData)
+
+		fastState := fast.NewInstrumentedState(state, oracle, os.Stdout, os.Stderr)
+		stepWitness, err := fastState.Step(true)
+		require.NoError(t, err)
+		require.False(t, stepWitness.HasPreimage())
+
+		require.Equal(t, pc+4, state.PC) // PC must advance
+		require.Equal(t, uint64(0), state.Heap)
+		require.Equal(t, uint64(0), state.LoadReservation)
+		require.Equal(t, uint8(0), state.ExitCode)
+		require.Equal(t, false, state.Exited)
+		require.Equal(t, preStateRoot, state.Memory.MerkleRoot())
+		require.Equal(t, step+1, state.Step) // Step must advance
+		require.Equal(t, preStatePreimageKey, state.PreimageKey)
+		require.Equal(t, expectedRegisters, state.Registers)
+
+		fastPost := state.EncodeWitness()
+		runEVM(t, contracts, addrs, stepWitness, fastPost)
+		runSlow(t, stepWitness, fastPost, oracle)
+	})
+}
+
+func FuzzStatePreimageWrite(f *testing.F) {
+	contracts := testContracts(f)
+	addrs := testAddrs
+
+	f.Fuzz(func(t *testing.T, addr uint64, count uint64, preimageOffset uint64, pc uint64, step uint64) {
+		pc = pc & 0xFF_FF_FF_FF_FF_FF_FF_FC // align PC
+		preimageData := []byte("hello world")
+		if preimageOffset >= uint64(len(preimageData)) {
+			t.SkipNow()
+		}
+		state := &fast.VMState{
+			PC:              pc,
+			Heap:            0,
+			ExitCode:        0,
+			Exited:          false,
+			Memory:          fast.NewMemory(),
+			LoadReservation: 0,
+			Registers:       [32]uint64{17: fast.SysWrite, 10: fast.FdPreimageWrite, 11: addr, 12: count},
+			Step:            step,
+			PreimageKey:     preimage.Keccak256Key(crypto.Keccak256Hash(preimageData)).PreimageKey(),
+			PreimageOffset:  preimageOffset,
+		}
+		state.Memory.SetUnaligned(pc, syscallInsn)
+		preStateRoot := state.Memory.MerkleRoot()
+		expectedRegisters := state.Registers
+		maxData := 32 - (addr & 31)
+		if maxData < count {
+			count = maxData
+		}
+		expectedRegisters[10] = count
+		expectedRegisters[11] = 0
+
+		oracle := staticOracle(t, preimageData)
+
+		fastState := fast.NewInstrumentedState(state, oracle, os.Stdout, os.Stderr)
+		stepWitness, err := fastState.Step(true)
+		require.NoError(t, err)
+		require.False(t, stepWitness.HasPreimage())
+
+		require.Equal(t, pc+4, state.PC) // PC must advance
+		require.Equal(t, uint64(0), state.Heap)
+		require.Equal(t, uint64(0), state.LoadReservation)
+		require.Equal(t, uint8(0), state.ExitCode)
+		require.Equal(t, false, state.Exited)
+		require.Equal(t, preStateRoot, state.Memory.MerkleRoot())
+		require.Equal(t, step+1, state.Step) // Step must advance
+		require.Equal(t, uint64(0), state.PreimageOffset)
```

Comment:
AFAIK, Preimage write sets state's PreimageKey using preimagedata. We should check this field is set as well, after stepping.

Reply:
Fixed preimage write test to verify functionality better